.travis.yml

```diff
@@ -1,81 +1,115 @@
+
 jobs:
   include:
+    ## Build cli, Psiphon ConsoleClient and Android app
+    # Go versions: first value in list
     - &build # YAML anchor/alias
       stage: build
       name: "Build cli and Psiphon ConsoleClient on Linux"
+      env:
+        - STAGE_BUILD=1 # Force separate caches
       install:
-        - go get ./...
+        - go get -d -u ./...
         # Substitute build string
         - sed -i.bak "s/\[BUILD\]/$TRAVIS_BRANCH-$TRAVIS_COMMIT/" tapdance/logger.go
-        - mkdir -p $GOPATH/src/github.com/Psiphon-Labs
-        - git clone https://github.com/Psiphon-Labs/psiphon-tunnel-core.git $GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core
+        - if [ -d "$GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core" ]; then (cd $GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core && git reset --hard); fi
+        - go get -d -u github.com/Psiphon-Labs/psiphon-tunnel-core/psiphon
         # Remove gotapdance from vendored packages
-        - go get github.com/kardianos/govendor
+        - go get -u github.com/kardianos/govendor
         - (cd $GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core && govendor remove github.com/sergeyfrolov/gotapdance/...)
         # Enable TapDance logging
-        - sed -i.bak 's/refraction_networking_tapdance.Logger().Out = ioutil.Discard//' $GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core/psiphon/common/tapdance/tapdance.go
+        - sed -i.bak "s/refraction_networking_tapdance\.Logger()\.Out = ioutil\.Discard//" $GOPATH/src/github.com/Psiphon-Labs/psiphon-tunnel-core/psiphon/common/tapdance/tapdance.go
       script:
         - go build -o build/cli-$TRAVIS_OS_NAME ./cli
-        - go build -o build/ConsoleClient-$TRAVIS_OS_NAME -tags 'TAPDANCE' github.com/Psiphon-Labs/psiphon-tunnel-core/ConsoleClient
+        - go build -o build/ConsoleClient-$TRAVIS_OS_NAME -tags "TAPDANCE" github.com/Psiphon-Labs/psiphon-tunnel-core/ConsoleClient
       after_success:
         # Upload built binaries to S3
         - sudo pip install awscli
-        - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then aws s3 sync build s3://$ARTIFACTS_BUCKET/$TRAVIS_REPO_SLUG/$TRAVIS_BRANCH/; fi
+        - if [ "$TRAVIS_PULL_REQUEST" = "false" ]; then aws s3 sync build s3://$AWS_BUCKET/$TRAVIS_REPO_SLUG/$TRAVIS_BRANCH/; fi
 
     - <<: *build # Same build on OS X
       name: "Build cli and Psiphon ConsoleClient on OS X"
       os: osx
 
     - <<: *build # Reuse setup steps from build
       name: "Build Psiphon Android Library and App"
+      dist: trusty # xenial not available
+      language: android
+      android:
+        components:
+          - android-23
       before_script:
-        # Get Android build environment
-        - docker pull refraction/psiandroid
-        - mkdir -p $GOPATH/src/bitbucket.org/psiphon
-        - hg clone https://bitbucket.org/psiphon/psiphon-circumvention-system $GOPATH/src/bitbucket.org/psiphon/psiphon-circumvention-system
+        # Install SDK and accept licenses
```

Comment:
Psiphon's Dockerfile isn't ideal to begin with. I have already prebuilt their image on Docker Hub (https://hub.docker.com/r/refraction/psiandroid/) but pulling it (around 3GB) for every build still takes about 8 mins. 

Those patches are necessary to support their use of x/gomobile paradigms from 3yo commits that have since been deprecated, and are also present (arguably in an even uglier form) in the Dockerfile https://github.com/Psiphon-Labs/psiphon-tunnel-core/blob/master/MobileLibrary/Android/Dockerfile

Reply:
> Psiphon's Dockerfile isn't ideal to begin with.

That's not my point. I just want that to be their job, if possible.

> pulling it (around 3GB) for every build still takes about 8 mins.

Can we cache that?

Reply:
Unfortunately, Docker load times out Travis (> 10 mins), so no. That was one of the approaches I've tried

Reply:
You can't cache a Docker pull. The only ways to have a local copy (at least without advanced wizardry) is either (1) to store it in your Docker cache, or (2) tar it out via docker save/load